File: packages/gavio-py/gavio/interceptors/pii/scanners/email.py
"""Email address scanner (RFC 5322 pragmatic subset)."""

from __future__ import annotations

import re

from ..context import ScanContext
from ..match import PiiMatch
from ..scanner import PiiScanner

_EMAIL = re.compile(
    r"\b[A-Za-z0-9._%+\-]+@[A-Za-z0-9.\-]+\.[A-Za-z]{2,}\b"
)


class EmailScanner(PiiScanner):
    entity_type = "EMAIL"

    def scan(self, text: str, ctx: ScanContext) -> list[PiiMatch]:
        out: list[PiiMatch] = []
        for m in _EMAIL.finditer(text):
            idx = ctx.next_index(self.entity_type)
            out.append(
                PiiMatch(
                    entity_type=self.entity_type,
                    start=m.start(),
                    end=m.end(),
                    value=m.group(),
                    replacement=f"[EMAIL_{idx}]",
                )
            )
        return out

File: packages/gavio-js/src/interceptors/pii/scanners/email.ts
/** Email address scanner (RFC 5322 pragmatic subset). */

import type { ScanContext } from '../context.js'
import { makeMatch } from '../match.js'
import type { PiiMatch } from '../match.js'
import type { PiiScanner } from '../scanner.js'

const EMAIL = /[A-Za-z0-9._%+-]+@[A-Za-z0-9.-]+\.[A-Za-z]{2,}/g

export function emailScanner(): PiiScanner {
  return {
    entityType: 'EMAIL',
    tier: 1,
    scan(text: string, ctx: ScanContext): PiiMatch[] {
      const out: PiiMatch[] = []
      for (const m of text.matchAll(EMAIL)) {
        const idx = ctx.nextIndex('EMAIL')
        out.push(
          makeMatch({
            entityType: 'EMAIL',
            start: m.index,
            end: m.index + m[0].length,
            value: m[0],
            replacement: `[EMAIL_${idx}]`,
          }),
        )
      }
      return out
    },
  }
}
